nvaccess/nvda
# Expose transparent background colors for displayModel

### Link to issue number:
Resolves concerns about contributed PR: https://github.com/nvaccess/nvda/pull/9197
Workaround for #6467

### Summary of the issue:

From #6467 (comment)

>When text is written with the GDI background mode set to transparent, the display model doesn't necessary report the correct background color for that text. In this case, the GDI hook should look at the actual background color before the text is drawn, rather than using the result of GetBkColor.


### Description of how this pull request fixes the issue:
There have been two prior attempts to fix this, both causing severe performance regressions.
- https://github.com/nvaccess/nvda/pull/6844
- https://github.com/nvaccess/nvda/pull/7440

This implementation merely exposes information about the potential transparency of the text background.
[Testing with various GDI dependent GUI APIs](https://github.com/nvaccess/testDisplayModel) shows that the `GetBkColor` result can be:
- Opaque and matches visual presentation(rawGDI)
- Transparent and matches visual presentation: (mfcTest, GDItest)
- Transparent and does not match visual presentation: (annecdotal)

The exposure of the transparency (via the `alpha` value on the `colors.RGB` class) will allow appModules 
or add-ons to customize NVDA's presentation as necessary.

For users, there is no change in behavior by default. An option is introcuded to the advanced settings panel to allow reporting of transparency in colors. The intention here is that developers will use this to identify where
color reporting could be improved.

The prior approach used the colorref high-level byte of a `colorref` struct. However, MSDN requires this byte be zero.
This prior approach did not pass the colorref into any Windows API's, so it was safe. But to prevent this from being accidentally introduced in the future, a compatible alternative type is introduced and used instead.

In order to stay compatible with python code interpretting this color value, a single bit is set if the color is considered transparent.

### Testing strategy:
- Built applications from https://github.com/nvaccess/testDisplayModel
- Tested with/without transparency reporting enabled (advanced panel)
   - Ensure color reporting is enabled in document formatting panel of NVDA settings
   - Use `NVDA+Numpad 7` to switch to screen review
   - Use `Numpad 7`, `Numpad 8`, `Numpad 9` to read the dialog

### Known issues with pull request:
None

### Change log entries:
For Developers
```
- Transparency of text background color, sourced from GDI applications (via the display model) are now exposed for add-ons or appModules.
```

### Code Review Checklist:

  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
